src/main/java/org/numenta/nupic/encoders/RandomDistributedScalarEncoder.java

```diff
@@ -0,0 +1,676 @@
+	 * 
+	 * @param rep1
+	 * @param rep2
+	 * @return
+	 */
+	public int countOverlap(List<Integer> rep1, List<Integer> rep2) {
+		int overlap = 0;
+		for (int index : rep1) {
+			for (int index2 : rep2)
+				if (index == index2)
+					overlap++;
+		}
+		return overlap;
+	}
+
+	/**
+	 * Return the overlap between two representations. rep1 and rep2 are arrays
+	 * of non-zero indices.
+	 * 
+	 * @param rep1
+	 * @param rep2
+	 * @return
+	 */
+	public int countOverlap(int[] rep1, int[] rep2) {
+		int overlap = 0;
+		for (int index : rep1) {
+			for (int index2 : rep2)
+				if (index == index2)
+					overlap++;
+		}
+		return overlap;
+	}
+
+	/**
+	 * Return True if the given overlap between bucket indices i and j are
+	 * acceptable.
+	 * 
+	 * @param i
+	 * @param j
+	 * @param overlap
+	 * @return
+	 */
+	public boolean overlapOK(int i, int j, int overlap) {
+		if (Math.abs(i - j) < getW())
+		{
+			if((overlap == (getW() - Math.abs(i - j))))
+				return true;
+			else
+				return false;
+		}
+		else
+		{
+			if(overlap <= getMaxOverlap())
+				return true;
+			else
+				return false;
+		}
+	}
+
+	/**
+	 * Return True if the given overlap between bucket indices i and j are
+	 * acceptable. The overlap is calculate from the bucketMap.
+	 * @param i
+	 * @param j
+	 * @return
+	 * @throws IllegalStateException
+	 */
+	public boolean overlapOK(int i, int j) throws IllegalStateException {
+		return overlapOK(i, j, countOverlapIndices(i, j));
+	}
+
+	/**
+	 * Return the overlap between bucket indices i and j
+	 * 
+	 * @param i
+	 * @param j
+	 * @return
+	 * @throws IllegalStateException
+	 */
+	public int countOverlapIndices(int i, int j) throws IllegalStateException {
+		boolean containsI = bucketMap.containsKey(i);
+		boolean containsJ = bucketMap.containsKey(j);
+		if (containsI && containsJ) {
+			List<Integer> rep1 = bucketMap.get(i);
+			List<Integer> rep2 = bucketMap.get(j);
+			return countOverlap(rep1, rep2);
+		} else if (!containsI)
+			throw new IllegalStateException("index " + i + " don't exist");
+		else
+			throw new IllegalStateException("index " + j + " don't exist");
+	}
+
+	/**
+	 * Given a bucket index, return the list of non-zero bits. If the bucket
+	 * index does not exist, it is created. If the index falls outside our range
+	 * we clip it.
+	 * 
+	 * @param index 
+	 * @return The list of active bits in the representation
+	 * @throws IllegalStateException
+	 */
+	public List<Integer> mapBucketIndexToNonZeroBits(int index)
+			throws IllegalStateException {
+		if (index < 0)
+			index = 0;
+
+		if (index >= getMaxBuckets())
+			index = getMaxBuckets() - 1;
+
+		if (!bucketMap.containsKey(index)) {
+			if (getVerbosity() >= 2)
+				System.out
+						.println("Adding additional buckets to handle index= "
+								+ index);
+			createBucket(index);
+		}
+		return bucketMap.get(index);
+	}
+
+	@Override
+	public int[] getBucketIndices(double x) {
+		if (Double.isNaN(x))
+			x = Encoder.SENTINEL_VALUE_FOR_MISSING_DATA;
```

Comment:
I believe the purpose of this is so that in the future if someone decides to change the sentinel value to something else, the code will still run and change NaN to that value. I would still like to have someone else's perspective on this